tools/protoxform/protoxform.py

```diff
@@ -0,0 +1,410 @@
+# protoc plugin to map from FileDescriptorProtos to a canonicaly formatted
+# proto.
+#
+# protoxform is currently only a formatting tool, but will act as the basis for
+# vN -> v(N+1) API migration tooling, allowing for things like deprecated field
+# removal, package renaming, field movement, providing both .proto and .cc code
+# generation to support automation of Envoy API version translation.
+#
+# See https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto
+# for the underlying protos mentioned in this file. See
+
+from collections import deque
+import functools
+import os
+import re
+import subprocess
+
+from tools.api_proto_plugin import plugin
+from tools.api_proto_plugin import visitor
+
+from google.api import annotations_pb2
+from google.protobuf import text_format
+from validate import validate_pb2
+
+CLANG_FORMAT_STYLE = ('{ColumnLimit: 100, SpacesInContainerLiterals: false, '
+                      'AllowShortFunctionsOnASingleLine: false}')
+
+
+class ProtoXformError(Exception):
+  """Base error class for the protoxform module."""
+
+
+def ClangFormat(contents):
+  """Run proto-style oriented clang-format over given string.
+
+  Args:
+    contents: a string with proto contents.
+
+  Returns:
+    clang-formatted string
+  """
+  return subprocess.run(
+      ['clang-format',
+       '--style=%s' % CLANG_FORMAT_STYLE, '--assume-filename=.proto'],
+      input=contents.encode('utf-8'),
+      stdout=subprocess.PIPE).stdout
+
+
+def FormatBlock(block):
+  """Append \n to a .proto section (e.g.
+
+  comment, message definition, etc.) if non-empty.
+
+  Args:
+    block: a string representing the section.
+
+  Returns:
+    A string with appropriate whitespace.
+  """
+  if block.strip():
+    return block + '\n'
+  return ''
+
+
+def FormatComments(comments):
+  """Format a list of comment blocks from SourceCodeInfo.
+
+  Prefixes // to each line, separates blocks by spaces.
+
+  Args:
+    comments: a list of blocks, each block is a list of strings representing
+      lines in each block.
+
+  Returns:
+    A string reprenting the formatted comment blocks.
+  """
+
+  # TODO(htuch): not sure why this is needed, but clang-format does some weird
```

Comment:
I guess this is related to my other comment. Technically I think this makes the code in the comment incorrect bash, so would be nice to fix this somehow but not a crisis if we can't.

Reply:
Yeah, I don't get it; it's a `clang-format` annoyance, I didn't figure out what options could avoid this. I'll leave this one as is for now and take a TODO (with relatively low priority) to fix it one day.